File: bench/src/python/map_kernel.py
# Equal-workload kernel -- see bench/src/c/map_kernel.c. Same open-addressing map, same hash, same
# fixed prealloc, no stdlib dict (Python's `dict` is measured in the `map` row instead).
import sys

CAP = 262144          # > N / 0.75, power of 2
N, PASSES = 100000, 10

keys = [0] * CAP
vals = [0] * CAP
used = bytearray(CAP)


def slot_for(k):
    return ((k & 0xFFFFFFFF) * 2654435761) % CAP    # hash


def put(k, v):
    i = slot_for(k)
    while used[i]:
        if keys[i] == k:
            vals[i] = v
            return
        i += 1
        if i >= CAP:
            i = 0
    keys[i] = k
    vals[i] = v
    used[i] = 1


def get(k):
    i = slot_for(k)
    while used[i]:
        if keys[i] == k:
            return vals[i]
        i += 1
        if i >= CAP:
            i = 0
    return 0


for i in range(N):
    put(i, i * 2)
s = 0
for _ in range(PASSES):
    for i in range(N):
        s += get((i * 2654435761) % N)
sys.exit(s % 256)

File: bench/src/lua/map_kernel.lua
-- Equal-workload kernel -- see bench/src/c/map_kernel.c. Same open-addressing map, same hash, same
-- fixed prealloc, no stdlib table-as-map (Lua's native table is measured in the `map` row instead).
local CAP = 262144        -- > N / 0.75, power of 2
local N, PASSES = 100000, 10

local keys, vals, used = {}, {}, {}
for i = 0, CAP - 1 do used[i] = 0 end

local function slotFor(k)
    return ((k & 0xFFFFFFFF) * 2654435761) % CAP    -- hash (Lua 5.4 has 64-bit integers)
end

local function put(k, v)
    local i = slotFor(k)
    while used[i] ~= 0 do
        if keys[i] == k then vals[i] = v; return end
        i = i + 1; if i >= CAP then i = 0 end
    end
    keys[i] = k; vals[i] = v; used[i] = 1
end

local function get(k)
    local i = slotFor(k)
    while used[i] ~= 0 do
        if keys[i] == k then return vals[i] end
        i = i + 1; if i >= CAP then i = 0 end
    end
    return 0
end

for i = 0, N - 1 do put(i, i * 2) end
local sum = 0
for _ = 1, PASSES do
    for i = 0, N - 1 do
        sum = sum + get((i * 2654435761) % N)
    end
end
os.exit(sum % 256)
